File: android/src/main/java/com/lklima/video/manager/RNVideoManagerModule.java

package com.lklima.video.manager;

import com.coremedia.iso.boxes.Container;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;

import android.graphics.Bitmap;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.util.Log;
import com.facebook.react.bridge.ReadableArray;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableNativeMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.googlecode.mp4parser.authoring.Movie;
import com.googlecode.mp4parser.authoring.Track;
import com.googlecode.mp4parser.authoring.container.mp4.MovieCreator;
import com.googlecode.mp4parser.authoring.tracks.AppendTrack;
import com.googlecode.mp4parser.authoring.builder.DefaultMp4Builder;

import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.net.URI;
import java.nio.channels.FileChannel;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class RNVideoManagerModule extends ReactContextBaseJavaModule {

  private final ReactApplicationContext reactContext;

  public RNVideoManagerModule(ReactApplicationContext reactContext) {
    super(reactContext);
    this.reactContext = reactContext;
  }

  @ReactMethod
  public void getVideoMetadataFor(ReadableArray fileNames, Promise promise) {
    try {
      WritableMap metadata = new WritableNativeMap();

      for (int i = 0; i < fileNames.size(); i++) {
        String fileName = fileNames.getString(i);
        MediaMetadataRetriever retriever = new MediaMetadataRetriever();
        retriever.setDataSource(reactContext.getApplicationContext(), Uri.parse(fileName.replaceFirst("file://", "")));
        String time = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);

        if (time == null) {
          time = "-1000";
        }

        int seconds = (int) (Long.parseLong(time) / 1000);

        WritableMap fileMetadata = new WritableNativeMap();
        fileMetadata.putInt("duration", seconds);
        fileMetadata.putBoolean("playable", true);

        metadata.putMap(fileName, fileMetadata);
      }

      promise.resolve(metadata);
    } catch (IOException e) {
      throw new RuntimeException(e);
    } catch (Error e) {
      e.printStackTrace();
      promise.reject(e.getMessage());
    }
  }

  @ReactMethod
  public void generateThumbnailFor(String video, ReadableMap options, Promise promise) {
    String writeDirectory = options.getString("writeDirectory");
    String fileName = options.getString("fileName");
    long timestamp = (long) options.getDouble("timestamp");

    String filePath = String.format("%s/%s.png", writeDirectory, fileName);

    try {
      MediaMetadataRetriever retriever = new MediaMetadataRetriever();
      Bitmap thumbnailBitmap = null;

      retriever.setDataSource(reactContext.getApplicationContext(), Uri.parse(video.replaceFirst("file://", "")));

      thumbnailBitmap = retriever.getFrameAtTime(timestamp * 1000000, MediaMetadataRetriever.OPTION_CLOSEST_SYNC);

      try (FileOutputStream out = new FileOutputStream(filePath)) {
        thumbnailBitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
      } catch (IOException e) {
        e.printStackTrace();
      }

      promise.resolve(true);
    } catch (Error e) {
      e.printStackTrace();
      promise.reject(e.getMessage());
    }
  }

  @ReactMethod
  public void merge(ReadableArray videoFiles, Promise promise) {

    List<Movie> inMovies = new ArrayList<Movie>();

    for (int i = 0; i < videoFiles.size(); i++) {
     String videoUrl = videoFiles.getString(i).replaceFirst("file://", "");

      try {
        inMovies.add(MovieCreator.build(videoUrl));
      } catch (IOException e) {
        promise.reject(e.getMessage());
        e.printStackTrace();
      }
    }

    List<Track> videoTracks = new LinkedList<Track>();
    List<Track> audioTracks = new LinkedList<Track>();

    for (Movie m : inMovies) {
      for (Track t : m.getTracks()) {
        if (t.getHandler().equals("soun")) {
          audioTracks.add(t);
        }
        if (t.getHandler().equals("vide")) {
          videoTracks.add(t);
        }
      }
    }

    Movie result = new Movie();

    if (!audioTracks.isEmpty()) {
      try {
        result.addTrack(new AppendTrack(audioTracks.toArray(new Track[audioTracks.size()])));
      } catch (IOException e) {
        promise.reject(e.getMessage());
        e.printStackTrace();
      }
    }
    if (!videoTracks.isEmpty()) {
      try {
        result.addTrack(new AppendTrack(videoTracks.toArray(new Track[videoTracks.size()])));
      } catch (IOException e) {
        promise.reject(e.getMessage());
        e.printStackTrace();
      }
    }

    Container out = new DefaultMp4Builder().build(result);
    FileChannel fc = null;

    try {

      Long tsLong = System.currentTimeMillis()/1000;
      String ts = tsLong.toString();

      String outputVideo = reactContext.getApplicationContext().getCacheDir().getAbsolutePath()+"output_"+ts+".mp4";

      fc = new RandomAccessFile(String.format(outputVideo), "rw").getChannel();

      Log.d("VIDEO", String.valueOf(fc));
      out.writeContainer(fc);
      fc.close();
      promise.resolve(outputVideo);
    } catch (FileNotFoundException e) {
      e.printStackTrace();
      promise.reject(e.getMessage());
    } catch (IOException e) {
      e.printStackTrace();
      promise.reject(e.getMessage());
    }

  }

  @Override
  public String getName() {
    return "RNVideoManager";
  }
}

File: android/src/main/java/com/lklima/video/manager/RNVideoManagerModule.kt
package com.lklima.video.manager

import android.util.Log
import com.facebook.react.bridge.NativeMap
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableNativeMap
import com.facebook.react.module.annotations.ReactModule
import com.lklima.video.manager.merger.MergedVideoResults
import com.lklima.video.manager.merger.MergedVideoResultsMapFactory
import com.lklima.video.manager.merger.UriSanitizer
import com.lklima.video.manager.merger.VideoMerger
import com.lklima.video.manager.merger.ffmpeg.VideoMergerFfmpeg
import com.lklima.video.manager.merger.options.MergedVideoOptionsFactory
import com.lklima.video.manager.metadata.VideoMetadataExtractor
import com.lklima.video.manager.thumbnail.VideoThumbnailGenerator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@ReactModule(name = "RNVideoManager")
class RNVideoManagerModule(
    private val reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext) {

    private val scope = MainScope()

    private val mergedVideoOptionsFactory by lazy {
        MergedVideoOptionsFactory(reactContext)
    }

    private val mergedVideoResultsMapFactory by lazy {
        MergedVideoResultsMapFactory()
    }

    @ReactMethod
    fun getVideoMetadataFor(fileNames: ReadableArray, promise: Promise) {
        scope.launch(Dispatchers.Main) {
            val files = buildList {
                for (i in 0 until fileNames.size()) {
                    add(fileNames.getString(i))
                }
            }
            VideoMetadataExtractor(reactContext).getVideoMetadataFor(
                files = files,
            ).onSuccess { map: Map<String, Long> ->
                Log.v(TAG, "get video metadata success $map")
                val metadata = WritableNativeMap()
                map.forEach { (fileName: String, durationSecs: Long) ->
                    WritableNativeMap().apply {
                        putInt("duration", durationSecs.toInt())
                        putBoolean("playable", true)
                    }.let {
                        metadata.putMap(fileName, it)
                    }
                }
                Log.v(TAG, "get video metadata success $metadata")
                promise.resolve(metadata)
            }.onFailure { error ->
                Log.e(TAG, "get video metadata failure", error)
                promise.reject("failed to get video metadata ${error.message}", error)
            }
        }
    }

    @ReactMethod
    fun generateThumbnailFor(video: String, options: ReadableMap, promise: Promise) {
        scope.launch(Dispatchers.Main) {
            VideoThumbnailGenerator(reactContext).generateThumbnail(
                video = video,
                writeDirectory = requireNotNull(options.getString("writeDirectory")) {
                    "no write directory provided in options map"
                },
                fileName = requireNotNull(options.getString("fileName")) {
                    "no filename provided in options map"
                },
                timestamp = options.getDouble("timestamp").toLong()
            ).onSuccess { success ->
                Log.v(TAG, "generate thumbnail success? $success")
                promise.resolve(success)
            }.onFailure { error ->
                Log.e(TAG, "generate thumbnail failure", error)
                promise.reject("failed to generate thumbnail: ${error.message}", error)
            }
        }
    }

    @ReactMethod
    fun merge(videoFiles: ReadableArray, options: ReadableMap, promise: Promise) {
        scope.launch(Dispatchers.IO) {
            runCatching {
                val files = buildList {
                    for (index in 0 until videoFiles.size()) {
                        add(videoFiles.getString(index))
                    }
                }
                val uriSanitizer = UriSanitizer()
                files.map(uriSanitizer::sanitize)
            }.mapCatching { uris ->
                val videoMerger: VideoMerger = VideoMergerFfmpeg(reactContext)
                videoMerger.mergeVideos(
                    videoFiles = uris,
                    options = mergedVideoOptionsFactory.newInstance(options)
                ).getOrThrow()
            }.mapCatching { output: MergedVideoResults ->
                mergedVideoResultsMapFactory.newInstance(output)
            }.onSuccess { nativeMap: NativeMap ->
                Log.v(TAG, "merge success $nativeMap")
                withContext(Dispatchers.Main) {
                    promise.resolve(nativeMap)
                }
            }.onFailure { error ->
                Log.e(TAG, "merge failed", error)
                withContext(Dispatchers.Main) {
                    promise.reject(error)
                }
            }
        }
    }

    override fun getName(): String = "RNVideoManager"

    private companion object {
        const val TAG = "RNVideoManagerModule"
    }
}